### DISCUSSION TITLE
inferRouterOutputs mismatch in types backend shows Date and frontend shows String

### DISCUSSION THREAD

**Discussion by kaykhan** [trpc/trpc#5012] [Q&A] [answered]
I have a declared public procedure `getRegistry`. I want to create a type on the frontend that matches the output of this public procedure.

relevant documentation: https://trpc.io/docs/client/vanilla/infer-types#inferring-input--output-types

However i am noticing that the frontend infers the `Date` type as a `String`. I don't care what the type is i just would like for both of them to match. How do i do this?

**backend:**
```
export const appRouter = router({
    getRegistry: publicProcedure.query(async () => {
        const result = await db.select().from(registry);
        return result;
    }),
})
```

![image](https://github.com/trpc/trpc/assets/5202527/3122562f-ae91-42cb-9c51-e16b3c06b74c)

**frontend:**

```
import type { inferRouterInputs, inferRouterOutputs } from "@trpc/server";
import type { AppRouter } from "./index";

export type RouterInput = inferRouterInputs<AppRouter>;
export type RouterOutput = inferRouterOutputs<AppRouter>;
```

```
export type IRegistry = RouterOutput["getRegistry"];
```

![image](https://github.com/trpc/trpc/assets/5202527/fb87c25e-4f35-4c85-952a-f15ac774cba6)

**Context**

 this is within a next.js app, i have a frontend component that is expecting a property the property type is the output of getRegistry (IRegistry)

> Diagnostics:
> 1. Type '{ id: string; type: string; name: string; createdAt: Date; updatedAt: Date; deletedAt: Date | null; organisation: string; repository_name: string; url: string; latest_version: string | null; last_fetched: Date | null; }[]' is not assignable to type '{ id: string; type: string; name: string; createdAt: string; updatedAt: string; deletedAt: string | null; organisation: string; repository_name: string; url: string; latest_version: string | null; last_fetched: string | null; }[]'.
>      Type '{ id: string; type: string; name: string; createdAt: Date; updatedAt: Date; deletedAt: Date | null; organisation: string; repository_name: string; url: string; latest_version: string | null; last_fetched: Date | null; }' is not assignable to type '{ id: string; type: string; name: string; createdAt: string; updatedAt: string; deletedAt: string | null; organisation: string; repository_name: string; url: string; latest_version: string | null; last_fetched: string | null; }'.
>        Types of property 'createdAt' are incompatible.
>          Type 'Date' is not assignable to type 'string'. [2322]

**Comment by KATT** [comment, marked as answer]
If you don't use any [data transformer](https://trpc.io/docs/server/data-transformers), `inferRouterInput`/`inferRouterOutput` will infer the way you actually will receive them in the API (the equivalent of doing `JSON.parse(JSON.stringify(output))`)

**Reply by kaykhan** [reply]
> If you don't use any [data transformer](https://trpc.io/docs/server/data-transformers), `inferRouterInput`/`inferRouterOutput` will infer the way you actually will receive them in the API (the equivalent of doing `JSON.parse(JSON.stringify(output))`)

Ah okay thank you that makes sense, `JSON.parse(JSON.stringify(output)))` will convert the Date to a String. Which is why im seeing a String on the frontend. 

Just to be super clear you're suggesting i "should" be using a data transformer such as "superjson" to align the types?

**Reply by KATT** [reply]
Up to you, depends on what you are looking for

**Reply by kaykhan** [reply]
@KATT I am looking to align the types between the frontend and the backend so one does not show Date and the other showing String. I added SuperJSON as the transformer and it solved the issue and now both are Date. 

`Up to you, depends on what you are looking for`

Is there some alternative you are suggesting which you have in mind? 

As an aside if i wanted both of them to be String rather than Date, how might i do that?

**Comment by triptu** [comment]
An alternative to SuperJSON is to write a custom deserializer, especially if there is any pattern to the date fields. In my case, I didn't really care about any other fields than Dates, and the cost of SuperJSON was too high(often >50% of the total request time). Sample code -

```ts
/**
 * recursively go through all the date fields and cast them to Date objects.
 * JSON doesn't support dates natively, doing this at deserialization time
 * makes it easier to work with dates in the code. As long as a field ends with
 * "At" and is not null, it will be cast to a Date object
 */
function castDateFields<T>(payload: Record<string, any>): T {
  let key: keyof typeof payload;
  for (key in payload) {
    const value = payload[key];
    if (key.endsWith("At") && value) {
      try {
        payload[key] = new Date(value as string);
      } catch (error) {
        logger.warn(`Error parsing date - key=${key}, value=${value} error=${error}`);
      }
    } else if (Array.isArray(value)) {
      value.forEach((item) => {
        if (typeof item === "object") {
          castDateFields(item);
        }
      });
    } else if (typeof value === "object") {
      castDateFields(value);
    }
  }
  return payload as T;
}

const t = initTRPC
  .context<Context>()
  .meta<TRPCPanelMeta>()
  .create({
    transformer: {
      serialize: (data) => data,
      deserialize: (data) => {
        return  castDateFields(data);
      },
    },
  });
```

## Example of an extreme case

Before -
time - 250ms
<img width="785" alt="image" src="https://github.com/trpc/trpc/assets/16299282/d889883e-b1f0-48e7-ae38-4a437038cc0a">


After
time - 75ms
<img width="782" alt="image" src="https://github.com/trpc/trpc/assets/16299282/aba24d12-69bd-4038-a1a2-110d0204d4f5">

**Comment by marsilinou97** [comment]
Make sure `strictNullChecks` is true in your tsconfig and have the transformer configured (superjson or other)
